### PR TITLE
Removing cache keys from cache operations to improve performance

### PR DESCRIPTION
**Pull request by trwalke**
Fixes #
https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/issues/3393

**Changes proposed in this request**
Removing cache keys from cache operations.
Cache keys are now replaced with pre calculated strings.

**Testing**
Unit testing

**Performance impact**
1-3% increase in performance

**Documentation**
- [ ] All relevant documentation is updated.

Most important changes are locates in the items and keys folders
image

### CODE REVIEW

File: src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs [hunk 1]
```diff
@@ -118,6 +126,58 @@ internal MsalAccessTokenCacheItem WithExpiresOn(DateTimeOffset expiresOn)
...
+            CacheKey = MsalCacheKeys.GetCredentialKey(
+                HomeAccountId,
+                Environment,
+                _credentialDescriptor,
+                ClientId,
+                TenantId,
+                ScopeString,
+                _extraKeyParts);
+
+            InitiOSKey();
```
Comment (trwalke) [inline src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs:149]: Going to make iOS keys lazy in next commit

File: src/client/Microsoft.Identity.Client/Cache/Keys/MsalAccountCacheKeyData.cs [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+{
+    internal struct MsalAccountCacheKeyData
+    {
+        public string HomeAccountId { get; set; }
+
+        public string CacheKey { get; set; }
+
+        public IiOSKey iOSKey { get; set; }
+
+        public MsalAccountCacheKeyData(string homeAccountId, string cacheKey, IiOSKey iosKey)
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Keys/MsalAccountCacheKeyData.cs:20]: Probably needs some comments. What is this object and what is `cacheKey` ?

File: src/client/Microsoft.Identity.Client/Cache/Items/MsalAccountCacheItem.cs [hunk 3]
```diff
@@ -105,6 +107,11 @@ internal MsalAccountCacheItem()
         /// account ID in its token cache. Accounts with associated WAM account ID can be used in silent WAM flows.
         /// </summary>
         internal IDictionary<string, string> WamAccountIds { get; set; }
+        public string CacheKey { get; private set; }
+
+        public IiOSKey iOSCacheKey { get; private set; }
+
+        internal MsalAccountCacheKeyData MsalAccountCacheKeyData { get; private set; }
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Items/MsalAccountCacheItem.cs:114]: Is `MsalAccountCacheKeyData` object necessary? Can you get rid of it?

**Review by bgavrilMS — APPROVED** [review]
LGTM

File: src/client/Microsoft.Identity.Client/Cache/Keys/MsalAccountCacheKeyData.cs [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Cache.Keys
+{
+    internal struct MsalAccountCacheKeyData
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Keys/MsalAccountCacheKeyData.cs:12]: Maybe we can get rid of this? Just rely on HomeAccountId ?

File: src/client/Microsoft.Identity.Client/Cache/Keys/IosKey.cs [hunk 5]
```diff
@@ -0,0 +1,30 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+
+namespace Microsoft.Identity.Client.Cache.Keys
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Keys/IosKey.cs:10]: nit: consider deleting the "Keys" folder, there's just 1 item.

File: src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs [hunk 6]
```diff
@@ -66,6 +72,8 @@ internal class MsalAccessTokenCacheItem : MsalCredentialCacheItemBase
             RawClientInfo = rawClientInfo;
             HomeAccountId = homeAccountId;
             OboCacheKey = oboCacheKey;
+
+            InitCacheKey();
         }
 
         private MsalAccessTokenCacheItem(
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs:79]: No init cache key here?

File: src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs [hunk 7]
```diff
@@ -17,6 +18,9 @@ namespace Microsoft.Identity.Client.Cache.Items
 {
     internal class MsalAccessTokenCacheItem : MsalCredentialCacheItemBase
     {
+        private string[] _extraKeyParts;
+        private string _credentialDescriptor;
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs:22]: Isn't this a const for AccessToken?

File: src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs [hunk 8]
```diff
@@ -227,20 +294,9 @@ internal string ToJsonString()
...
-                Environment,
-                TenantId,
-                HomeAccountId,
-                ClientId,
-                ScopeString,
-                TokenType);
-        }
-
-        internal MsalIdTokenCacheKey GetIdTokenItemKey()
+        internal MsalIdTokenCacheItem GetIdTokenItemKey()
```
Comment (bgavrilMS) [inline src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs:297]: This method name says "get key" but it returns an actual object.

**Review by bgavrilMS — APPROVED** [review]

File: src/client/Microsoft.Identity.Client/Cache/Keys/MsalAccountCacheKeyData.cs (hunk 2 above)
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Cache/Keys/MsalAccountCacheKeyData.cs]: removed

File: src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs [hunk 9]
```diff
@@ -41,6 +45,8 @@ internal class MsalAccessTokenCacheItem : MsalCredentialCacheItemBase
             RawClientInfo = response.ClientInfo;
             HomeAccountId = homeAccountId;
             OboCacheKey = oboCacheKey;
+
+            InitCacheKey();
```
[the remark this answers is not in the document]
Comment (trwalke) [reply src/client/Microsoft.Identity.Client/Cache/Items/MsalAccessTokenCacheItem.cs:49]: only the final ctor sets the data needed for the `initkey` method so it needs to be in there.

**Review by pmaytak — APPROVED** [review]

**Review by SameerK-MSFT — APPROVED** [review]

**Pull request merged**